src/bin/stratis_dumpmetadata.rs

```diff
@@ -0,0 +1,44 @@
+use std::{
+    env,
+    fs::OpenOptions,
+    io::{Seek, SeekFrom},
+    process,
+    vec::Vec,
+};
+
+use serde_json::Value;
+
+use libstratis::engine::BDA;
+
+fn run(devpath: String) -> Result<(), String> {
+    let mut devfile = OpenOptions::new().read(true).open(&devpath).unwrap();
```

Comment:
Step (1)

Ok! Here's the first ```unwrap()``` in the function. We might as well address this one, because we'll quickly see results. We know, for example, that this will fail if the file just does not exist. Easy to test.

Recall that ```unwrap()``` can be called on either an ```Option``` or a ```Result``` type. We need to figure out which this one is. I know it's a ```Result```, so we'll take that as given.

Now, we decide on the message we want, that's up to you.

Our goal is to convert that ```unwrap()``` call to something that will ultimately have the effect we want: print a useful error message to stderr and then exit.

We can do that by converting that ```unwrap()``` call to something that will cause an immediate return from the ```run``` function with a useful error message if there was an error when opening the file.

We want to use the function ```map_err()```, see here: **https://doc.rust-lang.org/std/result/enum.Result.html#method.map_err.__**.

And we want to return immediately from the function if there is an error, so we use the ```?``` operator as well.

So, the call will look something like this:

```
.map_err(|the_io__err| <your chosen error string>)?;
```

instead of 

```
unwrap();
```

Reply:
Note

```
|the_io_error| <your chosen error string>
```

is just an anonymous function that you write that takes as an argument the io error from the open operation and returns a string. You see them here and there in Python; the keyword that identifies them in Python is ```lambda```. We have a few in the CLI.

Reply:
I'm getting a mess of errors when I try this.

First try:
```
error: expected type, found `"Device not found"`
  --> src/bin/stratis_dumpmetadata.rs:14:92
   |
14 |     let mut devfile = OpenOptions::new().read(true).open(&devpath).map_err(|the_io_error| <"Device not found">)?;
   |                                                                                            ^^^^^^^^^^^^^^^^^^ expected type
```

What does this mean?  If I remove the unwrap, I see this on the subsequent call of BDA::load:

```
error[E0277]: the trait bound `std::result::Result<std::fs::File, std::io::Error>: std::io::Read` is not satisfied
  --> src/bin/stratis_dumpmetadata.rs:16:25
   |
16 |     let bda = BDA::load(&mut devfile).unwrap().unwrap();
   |                         ^^^^^^^^^^^^ the trait `std::io::Read` is not implemented for `std::result::Result<std::fs::File, std::io::Error>`
```

Trying to enter that type doesn't work:
```
error: comparison operators cannot be chained
  --> src/bin/stratis_dumpmetadata.rs:14:104
   |
14 |     let mut devfile = OpenOptions::new().read(true).open(&devpath).map_err(|the_io_err| std::io::Error <"Device not found">)?;
   |                                                                                                        ^^^^^^^^^^^^^^^^^^^^
   |
   = help: use `::<...>` instead of `<...>` to specify type arguments
   = help: or use `(...)` if you meant to specify fn arguments
```

...at this point, I obviously did something so wrong, that rustc is giving me a non sequitur message, and I should just bail out here.